src/common/mod.rs

```diff
@@ -1,163 +1,169 @@
 }
 
-impl HardwareInfo {
-    pub fn new() -> Self {
-        let numa_mapping = DashMap::new();
-        let mut node = 0;
-        loop {
-            let path = format!("/sys/devices/system/node/node{}/cpulist", node);
-            if let Ok(f) = std::fs::File::open(path) {
-                let mut reader = std::io::BufReader::new(f);
-                let mut line = String::new();
-                if reader.read_line(&mut line).is_ok() {
-                    let ranges: Vec<&str> = line.trim().split(',').collect();
-                    for range in ranges {
-                        let parts: Vec<&str> = range.split('-').collect();
-                        if parts.len() == 1 {
-                            if let Ok(id) = parts[0].parse() {
-                                numa_mapping.insert(id, node);
-                            }
-                        } else if parts.len() == 2 {
-                            if let Ok(start) = parts[0].parse() {
-                                if let Ok(stop) = parts[1].parse() {
-                                    for id in start..=stop {
-                                        numa_mapping.insert(id, node);
-                                    }
-                                }
-                            }
-                        }
-                    }
-                }
-            } else {
-                break;
-            }
-            node += 1;
+impl Counter {
+    pub fn new(counter: &'static LazyCounter, heatmap: Option<&'static LazyHeatmap>) -> Self {
+        Self {
+            previous: None,
+            counter,
+            heatmap,
         }
-        Self { numa_mapping }
     }
-
-    pub fn get_numa(&self, core: u64) -> Option<u64> {
-        self.numa_mapping.get(&core).map(|v| *v.value())
+    pub fn set(&mut self, now: Instant, elapsed: f64, value: u64) {
+        if let Some(previous) = self.previous {
+            let delta = value.wrapping_sub(previous);
+            self.counter.add(delta);
+            if let Some(heatmap) = self.heatmap {
+                heatmap.increment(now, (delta as f64 / elapsed) as _, 1);
+            }
+        }
+        self.previous = Some(value);
     }
 }
 
-/// helper function to discover the number of hardware threads
-pub fn hardware_threads() -> Result<u64, ()> {
-    let path = "/sys/devices/system/cpu/present";
-    let f =
-        std::fs::File::open(path).map_err(|e| debug!("failed to open file ({:?}): {}", path, e))?;
-    let mut f = std::io::BufReader::new(f);
-
-    let mut line = String::new();
-    f.read_line(&mut line)
-        .map_err(|_| debug!("failed to read line"))?;
-    let line = line.trim();
-    let a: Vec<&str> = line.split('-').collect();
-    a.last()
-        .unwrap_or(&"0")
-        .parse::<u64>()
-        .map_err(|e| debug!("could not parse num cpus from file ({:?}): {}", path, e))
-        .map(|i| i + 1)
+#[macro_export]
+#[rustfmt::skip]
+macro_rules! counter {
```

Comment:
I'd like to see some documentations of `counter`, `heatmap`, `gauge`, and other combo types. There are also some considerations given to the range and granularity of the heatmap created, which would be good to document as well.